server/src/main/java/org/elasticsearch/index/translog/MultiSnapshot.java

```diff
@@ -31,24 +30,20 @@
  * A snapshot composed out of multiple snapshots
  */
 final class MultiSnapshot implements Translog.Snapshot {
-
     private final TranslogSnapshot[] translogs;
     private final int totalOperations;
-    private int overriddenOperations;
     private final Closeable onClose;
-    private int index;
-    private final SeqNoSet seenSeqNo;
+    private int index = 0;
+    private final LongLongHashMap seqNoToTerm; // for assertion purpose
 
     /**
      * Creates a new point in time snapshot of the given snapshots. Those snapshots are always iterated in-order.
      */
     MultiSnapshot(TranslogSnapshot[] translogs, Closeable onClose) {
         this.translogs = translogs;
         this.totalOperations = Arrays.stream(translogs).mapToInt(TranslogSnapshot::totalOperations).sum();
-        this.overriddenOperations = 0;
         this.onClose = onClose;
-        this.seenSeqNo = new SeqNoSet();
-        this.index = translogs.length - 1;
+        this.seqNoToTerm = Assertions.ENABLED ? new LongLongHashMap() : null;
```

Comment:
I wonder if we should rather check this in `assertConsistentHistoryBetweenTranslogAndLuceneIndex`